model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProvider.java

```diff
@@ -46,18 +52,15 @@ public void close() {
     @Override
     @SuppressWarnings("unchecked")
     public <V extends AbstractEntity, M> MapStorage<V, M> getStorage(Class<M> modelType, Flag... flags) {
-        factory.validateAndUpdateSchema(session, modelType);
-        return new MapStorage<V, M>() {
-            @Override
-            public MapKeycloakTransaction<V, M> createTransaction(KeycloakSession session) {
-                MapKeycloakTransaction<V, M> sessionTx = session.getAttribute(SESSION_TX_PREFIX + modelType.hashCode(), MapKeycloakTransaction.class);
-                if (sessionTx == null) {
-                    sessionTx = factory.createTransaction(modelType, em);
-                    session.setAttribute(SESSION_TX_PREFIX + modelType.hashCode(), sessionTx);
-                } 
-                return sessionTx;
-            }
-        };
+        // validate and update the schema for the storage.
+        this.factory.validateAndUpdateSchema(this.session, modelType);
+        // create the JPA transaction and enlist it if needed.
+        if (session.getAttribute(this.sessionTxKey) == null) {
+            KeycloakTransaction jpaTransaction = new JpaTransactionWrapper(em.getTransaction());
+            session.getTransactionManager().enlist(jpaTransaction);
+            session.setAttribute(this.sessionTxKey, jpaTransaction);
+        }
+        return session -> factory.createTransaction(modelType, em);
```

Comment:
Please use the full inline class form, `MapStorage` is intentionally not marked as `@FunctionalInterface`

Reply:
Just followed what we have in LDAP:
https://github.com/keycloak/keycloak/blob/a2d7cd7a5ca84f9eb1e08336c1381517c6d19b9d/model/map-ldap/src/main/java/org/keycloak/models/map/storage/ldap/LdapMapStorageProvider.java#L42-L49

Reply:
Thanks for spotting this one, I missed it there. Could you please file an issue for the LDAP and fix only the JPA code in this PR?
cc: @ahus1

Reply:
Done: https://github.com/keycloak/keycloak/issues/11373